addons/mail/static/src/models/thread/thread.js

```diff
@@ -129,6 +129,17 @@ function factory(dependencies) {
         // Public
         //----------------------------------------------------------------------
 
+        computeMessageUnreadCounter() {
+            let message_unread_counter = 0;
+            if (this.orderedMessages) {
+                let orderedMessages = !this.seen_message_id
+                    ? this.orderedMessages
+                    : this.orderedMessages.filter(message => message.id > this.seen_message_id);
+                message_unread_counter = orderedMessages.length;
+            }
+            this.update({ message_unread_counter });
+        }
```

Comment:
Not correct, unread counter is a "complete" info, while `orderedMessages` are usually incomplete. This means you could have fetched 30 messages and have an unread counter of 1000. This so-called "compute" would set 30 instead of 1000.

That's why the message_unread_counter must not be computed in JS. The design of incrementing/decrementing it based on flow kind of work 90% of the time but that's not good neither. The correct solution consists of server sending the new message_unread_counter.

Some mail/messaging client apps tend to keep counter until user makes a manual fetch, so this hints at how they tackle this issue (which is to trust server data, even when mismatch with [incomplete-by-design] local data at certain times).

Reply:
Base value for `message_unread_counter` should always come from the server. It should be coupled with the last message id that was used to compute it: the last existing and potentially unseen message at that time, not the last seen message.

Then the JS will know what messages were taken into account or not based on that id, and can increment a local counter accordingly.

3 variables need to be stored for this, the server value, the messaging if that was used for it, and the local value.

Reply:
I really don't understand why so much trouble is needed, the server doesn't return the unread counter each time it changes ? If yes, why don't the JS just wait for counter to be updated instead of computing it with so much complex stuff ?

Reply:
> I really don't understand why so much trouble is needed, the server doesn't return the unread counter each time it changes ? If yes, why don't the JS just wait for counter to be updated instead of computing it with so much complex stuff ?

Well, the JS just get unread counter initially or from a payload that comes from `channel_info()`, but other times it just increments/decrements based on business flow. This is a logic that is also used with other counters, like needaction counter. This is flawed, and the correct way should have been to send `message_unread_counter` as frequent as possible because this is a "complete" data, whereas most data in web client are "incomplete" by design.

I think this is a big flaw from the previous code that we become more aware with experience we have now. This is not trivial, because we have to find a balance of correctness (server sends unread counter as frequent as possible) and reactivity (web client deduces counter change before acknowledgement from server).

Reply:
After some experimentation, I think we should always ask the server to return it when it changes.
There is actually two flows where it is changed : a message is seen or a new message is received.
Couldn't we update the content of these notifications to contain the counter ?

If it's not possible then I'm struggling to determine how to compute the local counter. Anyhow I turn this, I fallback to compute on message_seen_id and orderedMessages, the last message id used by the server to compute server message unread counter is just used for initialization. And the computing is basically "count all messages with id > than message_seen_id" like it was before. So I wonder why it's needed to change what is done as it's exactly the same solution but with 3 variables instead of 1...

Reply:
(I know that it's not exactly counting all message not seen in the interface as you need to also add initial value of message_unread_counter, just forgot to mention it)

Reply:
Let's take an example.
A channel with 2000 messages in database and channel_info gives us 
```js
message_unread_counter: 40,
seen_message_id: 1960,
```

Then you fetch the first page of the channel so you have in store 30 unread messages which are in orderedMessages.
Let's say you saw a message so that 10 messages are still not seen, then `seen_message_id` should be updated to something like 1990 but `message_unread_counter` would stay at 40 but the local id has to be 10. So what's the point of the server values in here .. ?

Reply:
We have to find a balance, here it should decrement counter to `10`, but if the server ever sends another different value like `12` (2 new messages, which we could hypothetically imagine we haven't fetched them), then it must display `12`, not `10`.

The server is the source of truth, but it's slow to fetch, and we can't realistically fetch at a huge frequency neither... And counter must be perceived as instantaneous from user point of view (mostly when it's about recent messages).

Note that this looks very similar to approach for chat window fold state and channel pinned. This will likely be a behaviour that we want a wide range of fields to use, so we will have to consider a way to generalize this behaviour in an easy way, without having to manually managing pending/local/server fields.

Reply:
So if I understand well, the computing is not changed but we have to erase any computation made in js if server decide to update the value, right ? Then we could use a computed field depending on server value. The computed field has the server value if the seen message id is the same as the one used by server to compute its counter, in all other cases local counter is the amount of fetched messages with an id > seen_message_id. Couldn't we ?+

Reply:
Honestly, I don't know precisely how it should work in every cases possible. Main purpose of my comments were to give some points to reflect on. In some ways, I'm not entirely sure whether fold or pinned are robust as they are now... ...Maybe, but considering the amount of code they need and how it has to manually manage local/pending/server, the likelihood of errors are quite high. The more fields that need to act like the, the more likely we will have subtle bugs everywhere.

For the time being, we can't do much better than do most stuffs by hand, and make assumptions based on existing flows. This means the scenario I gave with `40` becoming `12` by server does not happen with current code, so you could omit it (it's just sad we can't manage such case, this shows client-server communications can easily break UI, so making a super dynamic and mostly consistent UI is complicated right now).

Reply:
Same example.
So, imagine we have 40 messages : 
```
message #n
message #n-1
....
message #n-29
....
message #n-39
...
```

Last seen message id `n-39`.

Let's initialize messaging and display the thread.
We have :
```
orderedMessages <- [message n-29, ..., message n]
lastMessage <- message n
message_unread_counter <- 40
lastKnownMessageToComputeMessageUnreadCounter <- n
seen_message_id <- n-39
```

In this case we should display 40 as `lastKnownMessageToComputeMessageUnreadCounter == lastMessage`.

Imagine now we received a `message_seen` notification for `message n-38`.
We have :
```
orderedMessages <- [message n-29, ..., message n]
lastMessage <- message n
message_unread_counter <- 40
lastKnownMessageToComputeMessageUnreadCounter <- n
seen_message_id <- n-38
```

You smelled it, nothing changed except  `seen_message_id` but it's still an unknown message and the situation here is the same as the previous one I mention. So we just can rely on `message_unread_counter` which is 40 and yet we should display 39. I don't know if it can happen in reality as we could only mark as seen a message that is already in store but if it can this approach can be a problem. We could avoid this by setting a max on the shown counter to be <= to the amount of message in store (and display something like 30+ instead of 40 for example)

Maybe another approach could be good : when we get message_unread_counter from server, whatever happen we copy that value to local value. And instead of recomputing the local value based on what is shown etc etc we update this value in each flow needing it (new message and message_seen notifications). That is simpler I think and is slightly the same as what was done before but we keep here the value received from server untouched by JS and we assume it is always correct when we receive it.

What do you guys think @alexkuhn & @seb-odoo ?

Reply:
This example show how problematic it is for JS to know for sure the counter. And this is obvious because JS is incomplete by design, while server is not.

Unfortunately discuss code (JS before refactoring & python) implemented things in a naïve way, with increment/decrement/fetch more/fetch less/etc. The web client receives an initial messaging state (`mail/init_messaging`) and everything afterwards are "diffs" (that the JS has to guess) that have to be 100% correct...

Lots of "guesses" has been eliminated thanks to refactoring that introduced models and fields + embraced incompleteness design. However, python code is barely changed and it needs to be attentive of JS struggles and be more cooperative as a result, such as providing more information to help JS not making guesses.

The case of `message_unread_counter` could mostly be eliminated if the server was just regularly sending the `message_unread_counter` (like as an acknowledgment in payload of a seen notification for instance). This would ensure that counter are mostly correct, and in bad cases the JS may do wrong guesses but at least it will be eventually be good.
 
It's still nice to have a reactive UI and not have to wait for network + server time, which is why it's good to still let JS make adjustments of counter itself.

#

So, sorry for putting a long comment with no precise solution for this PR... I see design issues of non-cooperative enough python code, and also lack of auto-support for python sync fields (i.e. an easy way to define such a field, not the boilerplate code of having 3 fields that we have to fully manage manually and use a bunch of heuristics...).

I clearly see a combination of both of your suggested approaches at the same time:

- dynamic counter adjustments in JS (e.g. instantaneously increment/decrement counter, wouldn't even mind change like from 40 to 30+ when we can't accurately guess, even though this shows server should provide some help).
- exact counter set in JS (e.g. on receiving channel seen notification, it also contains the server unread counter).

By the way... to be fair, it's kinda wrong to assume server value is the source of truth... It's a value at a certain point in time, and there could be a delay when server computes and sends to web client, and when web client itself processes it. So a robust system should let server frequently send complete information to JS with timestamp, and JS may have to make adjustments on these complete information (like it just marked a message as read after the timestamp... so it should decrement the server message unread counter once!).

Reply:
It's totally right and I think the solution is what @seb-odoo proposed : each time message_unread_counter is sent by the server, it should also send the last message it took into account when computing it, that's what I added here ('last_known_message_id") and then JS should handle this value correctly.

At the end, obviously, we need the python server to be more cooperative, including the fact to send that counter more often but do we want to do that right now or do we reschedule that part ? (based on the idea that we want to refactor JS without refactoring python)

What do you think we need to still do for the current PR to be acceptable ?

Reply:
Making python more cooperative in case of `message_unread_counter` looks too much for 13.4 as a fix.
Would do that in master instead at a later date.

Also the clean way would need something like task-2321017 (server-sync fields), but we don't know yet how to achieve this... And there's a moderately important bug to fix with this PR here. The current solution is fragile, but hopeful good enough on a short-term. Especially considering the machinery in place is not too cumbersome, as the complexity is just concentrated on computing `localMessageUnreadCounter`.

I'll do another pass of review, but the current approach here will be kept in the scope of this PR.

Reply:
Please create a task with the summary of this discussion (and then resolve here).

Reply:
https://www.odoo.com/web#id=2336349&action=333&active_id=1750&model=project.task&view_type=form&cids=1&menu_id=4720